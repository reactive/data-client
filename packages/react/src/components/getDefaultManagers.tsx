import type { DevToolsConfig, Manager } from '@data-client/core';

import {
  NetworkManager,
  SubscriptionManager,
  PollingSubscription,
  DevToolsManager,
} from '../managers/index.js';

/* istanbul ignore next */
/** Returns the default Managers used by DataProvider.
 *
 * @see https://dataclient.io/docs/api/getDefaultManagers
 */
let getDefaultManagers: ({
  devToolsManager,
  networkManager,
  subscriptionManager,
}?: GetManagersOptions) => Manager[] = ({
  networkManager,
  subscriptionManager = PollingSubscription,
} = {}) =>
  subscriptionManager === null ?
    [constructManager(NetworkManager, networkManager)]
  : [
      constructManager(NetworkManager, networkManager),
      constructManager(SubscriptionManager, subscriptionManager),
    ];
/* istanbul ignore else */
if (process.env.NODE_ENV !== 'production') {
  getDefaultManagers = ({
    devToolsManager,
    networkManager,
    subscriptionManager = PollingSubscription,
  }: GetManagersOptions = {}): Manager[] => {
    if (networkManager === null) {
      console.error('Disabling NetworkManager is not allowed.');
      networkManager = {};
    }
    const nm = constructManager(NetworkManager, networkManager);
    const managers: Manager[] = [nm];
    if (subscriptionManager !== null) {
      managers.push(constructManager(SubscriptionManager, subscriptionManager));
    }
    if (devToolsManager !== null) {
      managers.unshift(
        devToolsManager instanceof DevToolsManager ? devToolsManager : (
          new DevToolsManager(devToolsManager, nm.skipLogging.bind(nm))
        ),
      );
    }
    return managers;
  };
}
export { getDefaultManagers };

function constructManager<
  M extends { new (...args: any): Manager },
  O extends InstanceType<M> | ConstructorArgs<M>,
>(Mgr: M, optionOrInstance: O): InstanceType<M> {
  return optionOrInstance instanceof Mgr ?
      (optionOrInstance as any)
    : (new Mgr(optionOrInstance) as any);
}

export type GetManagersOptions = {
  devToolsManager?: DevToolsManager | DevToolsConfig | null;
  networkManager?: NetworkManager | ConstructorArgs<typeof NetworkManager>;
  subscriptionManager?:
    | SubscriptionManager
    | ConstructorArgs<typeof SubscriptionManager>
    | null;
};

export type ConstructorArgs<T extends { new (...args: any): any }> =
  T extends { new (options: infer O): any } ? O : never;
