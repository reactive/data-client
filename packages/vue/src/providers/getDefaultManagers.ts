import type { DevToolsConfig, Manager } from '@data-client/core';
import {
  NetworkManager,
  SubscriptionManager,
  PollingSubscription,
  DevToolsManager,
} from '@data-client/core';

/* istanbul ignore next */
/** Returns the default Managers used by DataProvider. */
let getDefaultManagers: (options?: GetManagersOptions) => Manager[] = (
  options = {},
) => {
  const { networkManager, subscriptionManager = PollingSubscription } = options;
  if (subscriptionManager === null) {
    return [constructManager(NetworkManager, networkManager ?? ({} as any))];
  }
  return [
    constructManager(NetworkManager, networkManager ?? ({} as any)),
    constructManager(SubscriptionManager, subscriptionManager),
  ];
};
/* istanbul ignore else */
if (process.env.NODE_ENV !== 'production') {
  getDefaultManagers = (options: GetManagersOptions = {}): Manager[] => {
    const {
      devToolsManager,
      networkManager,
      subscriptionManager = PollingSubscription,
    } = options;
    if (networkManager === null) {
      console.error('Disabling NetworkManager is not allowed.');
      // fall back to default options
    }
    const nm = constructManager(NetworkManager, networkManager ?? ({} as any));
    const managers: Manager[] = [nm];
    if (subscriptionManager !== null) {
      managers.push(constructManager(SubscriptionManager, subscriptionManager));
    }
    if (devToolsManager !== null) {
      let dtm: DevToolsManager;
      if (devToolsManager instanceof DevToolsManager) {
        dtm = devToolsManager;
      } else {
        dtm = new DevToolsManager(
          devToolsManager as any,
          nm.skipLogging.bind(nm),
        );
      }
      managers.unshift(dtm);
    }
    return managers;
  };
}
export { getDefaultManagers };

function constructManager<M extends { new (...args: any): Manager }>(
  Mgr: M,
  optionOrInstance: InstanceType<M> | ConstructorArgs<M>,
): InstanceType<M> {
  if (optionOrInstance instanceof Mgr) {
    return optionOrInstance as InstanceType<M>;
  }
  return new Mgr(optionOrInstance as any) as InstanceType<M>;
}

export type GetManagersOptions = {
  devToolsManager?: DevToolsManager | DevToolsConfig | null;
  networkManager?:
    | NetworkManager
    | ConstructorArgs<typeof NetworkManager>
    | null;
  subscriptionManager?:
    | SubscriptionManager
    | ConstructorArgs<typeof SubscriptionManager>
    | null;
};

export type ConstructorArgs<T extends { new (...args: any): any }> =
  T extends new (options: infer O) => any ? O : never;
