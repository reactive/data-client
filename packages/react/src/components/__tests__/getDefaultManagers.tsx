import {
  NetworkManager,
  SubscriptionManager,
  DevToolsManager,
} from '@data-client/core';

import { getDefaultManagers } from '../getDefaultManagers';

describe('getDefaultManagers()', () => {
  let warnspy: jest.Spied<any>;
  let debugspy: jest.Spied<any>;
  beforeEach(() => {
    warnspy = jest.spyOn(global.console, 'warn').mockImplementation(() => {});
    debugspy = jest.spyOn(global.console, 'info').mockImplementation(() => {});
  });
  afterEach(() => {
    warnspy.mockRestore();
    debugspy.mockRestore();
  });

  let errorSpy: jest.Spied<typeof console.error>;
  afterEach(() => {
    errorSpy.mockRestore();
  });
  beforeEach(() => {
    errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {}); // Suppress console.error
  });
  it('should have SubscriptionManager in default managers', () => {
    const subManagers = getDefaultManagers().find(
      manager => manager instanceof SubscriptionManager,
    );
    expect(subManagers).toBeDefined();
  });
  it('should have NetworkManager in default managers', () => {
    const networkManager = getDefaultManagers().find(
      manager => manager instanceof NetworkManager,
    );
    expect(networkManager).toBeDefined();
  });
  it('should have DevToolsManager in default managers', () => {
    const devtoolsMgr = getDefaultManagers().find(
      manager => manager instanceof DevToolsManager,
    );
    expect(devtoolsMgr).toBeDefined();
  });
  it('null option should disable a manager', () => {
    expect(
      getDefaultManagers({ devToolsManager: null }).find(
        manager => manager instanceof DevToolsManager,
      ),
    ).toBeUndefined();
    expect(
      getDefaultManagers({ subscriptionManager: null }).find(
        manager => manager instanceof SubscriptionManager,
      ),
    ).toBeUndefined();
  });

  it('networkManager should not be removable', () => {
    expect(
      // @ts-expect-error
      getDefaultManagers({ networkManager: null }).find(
        manager => manager instanceof NetworkManager,
      ),
    ).toBeDefined();
    expect(errorSpy.mock.calls).toMatchSnapshot();
  });

  it('manager constructor options should work', () => {
    const managers = getDefaultManagers({
      networkManager: { dataExpiryLength: 1 },
      devToolsManager: {
        maxAge: 1000,
      },
    });
    expect(
      managers.find(manager => manager instanceof NetworkManager)
        ?.dataExpiryLength,
    ).toBe(1);
    // check that the value set is not default
    expect(
      managers.find(manager => manager instanceof DevToolsManager)
        ?.maxBufferLength,
    ).not.toBe(DevToolsManager.prototype.maxBufferLength);
  });

  it('manager instance should work', () => {
    class MyDevTool extends DevToolsManager {
      maxBufferLength = 500;
    }
    const managers = getDefaultManagers({
      networkManager: new NetworkManager({ dataExpiryLength: 1 }),
      devToolsManager: new MyDevTool(),
    });
    expect(
      managers.find(manager => manager instanceof NetworkManager)
        ?.dataExpiryLength,
    ).toBe(1);
    // check that the value set is not default
    expect(
      managers.find(manager => manager instanceof DevToolsManager)
        ?.maxBufferLength,
    ).not.toBe(DevToolsManager.prototype.maxBufferLength);
  });
});
