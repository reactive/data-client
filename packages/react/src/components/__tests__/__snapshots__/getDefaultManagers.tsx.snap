// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getDefaultManagers() networkManager should not be removable 1`] = `
[
  [
    "Disabling NetworkManager is not allowed.",
  ],
]
`;
